feat: improved initialization and custom SimConnect events for Autobrake

aguther opened the pull request:
## Summary of Changes

### Improved initialization

This PR adds an improved initialization logic Rust and C++. For that purpose, a new local variable is introduced that indicates that the JavaScript code is fully initialized, and the user has pressed the _READY TO FLY_ button. The local variable is named `A32NX_IS_READY` and will contain the value `1` when the condition is true.

#### Usage in Rust
The following changes have been made:
- added simulation time and new ready signal to update context
- added functions in update context to ask if sim is ready
- allow to set initial condition for DelayedTrueLogicGate and DelayedPulseTrueLogicGate
- initialize gates of Autobrake system based on start state (on ground / in flight)
- only evaluate state and deploy Ram-Air-Turbine when sim is ready

#### Usage in C++
The following changes have been made:
- use new read signal to better initialize FCU; this should prevent seeing speed managed with target 0
- depending on the start state, the FCU gets initialized with different values:
  - C&D: use default power-on values (SPD = 100, ALT = 100)
  - Runway: at best guess ready values (SPD = 150, HDG = current heading, ALT = 15000)
  - In-Flight: near values + AP engage + ATHR engage (SPD = managed, HDG = current heading rounded on 10, ALT = target current altitude rounded to next 1000, VS = +/- 1000 fpm towards target altitude)

### New custom events

#### A32NX.AUTOBRAKE_SET
This PR adds a custom event `A32NX.AUTOBRAKE_SET` that can be invoked via SimConnect and it allows to set the Autobrake to a distinct state by using a parameter value (without the risk to deactivate it on second trigger).

The following values are supported:
- 0 = disarm
- 1 = LOW
- 2 = MED
- 3 = MAX

Max is only accepted when this is possible with the button pressed in the cockpit.

#### A32NX.AUTOBRAKE_SET_DISARM
This event set the state to disarm.

#### A32NX.AUTOBRAKE_SET_LO
This event set the state to LO.

#### A32NX.AUTOBRAKE_SET_MED
This event set the state to MED.

#### A32NX.AUTOBRAKE_SET_MAX
This event set the state to MAX.

#### A32NX.AUTOBRAKE_BUTTON_LO
This event is as if pressing the button `LO` in the cockpit. The advantage compared to the default event is that this button can be pressed much faster after each other (100 ms vs. 1500 ms).

#### A32NX.AUTOBRAKE_BUTTON_MED
This event is as if pressing the button `MED` in the cockpit. The advantage compared to the default event is that this button can be pressed much faster after each other (100 ms vs. 1500 ms).

#### A32NX.AUTOBRAKE_BUTTON_MAX
This event is as if pressing the button `MAX` in the cockpit. The advantage compared to the default event is that this button can be pressed much faster after each other (100 ms vs. 1500 ms).

### Extended disarm logic
When for some reason the MAX state is activated in flight, it is now ensured that this state is disarmed. There was no such check before.

## Additional context
Why is this PR introducing a new local variable and a new event?

On the development version (before this PR), the available local variable can be set, and the autobrake system is also accepting this as its own state. The drawback of this solution is that the MAX state is also accepted when it should not be available.

In my first try I added code that reset the MAX state in cases where this is not allowed and this worked in principle, but on the PFD in many cases `BRK MAX` was shown for a moment.

The reason is that the up to now existing local variable was used for setting the state but is also the state for all other components. So, before the Autobrake system was able to adapt to the allowed state it was already sent to all other reading components like the PFD.

The reason for a distinct event is to have a single event that can be used to set any state for the Autobrake system via SimConnect.

The requirement for more advanced hardware setups (that want the button behavior) is also fulfilled, because in those cases I anyway expect that people have tools like FSUIPC or SPAD.neXt and therefore can easily use the LVARs or custom events for the buttons.

## Testing instructions
- test if the Autobrake system can be controlled with the default events as expected and described
- test if the Autobrake system can be controlled with the new custom event as described
- test if the Autobrake system is working as expected

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

tracernz commented:
Does this work with the TCA? I would add does this work with the TCA autobrake binds to test instructions, as that was one of the main issues with getting the previous implementation right.

aguther commented:
I have tested it with the default events. Will retest with the default binding directly in MSFS to be sure.

aguther commented:
Ok, I did some tests using the default TCA configuration and can tell the following:

## Behavior before this PR

Default events are constantly fired by MSFS with the exception that events are not fired when the right mouse button is pressed to change the view; this seems not to apply in all views but in most. When an event is fired (i.e. `AUTOBRAKE_HI_SET`), this event sets the corresponding button LVAR (i.e. `OVHD_AUTOBRK_MAX_ON_IS_PRESSED`) to `1` for at least 1.5 s. This means the buttons are usually permanently set to `1` when on the TCA the hardware is for example set to HI. When the LVAR stays on `1`, then this state is ignored, because the press event is only detected when an LVAR goes from `0` to `1`. Now when you press the right mouse button long enough, the LVAR goes back to `0` and when you release the right mouse button it goes back to `1` which then disables the Autobrake system (or enables it again when you do the same again). Due to this overall behavior, you can still operate the buttons even when the TCA hardware button is for example on `HI`.

🟢 buttons can be operated when hardware button is set to a certain position
🟡 hardware and state can get out of sync
🔴 when using the right mouse button to change the view Autobrake state can change unintentionally

## New behavior

Default events are constantly fired by MSFS with the exception that events are not fired when the right mouse button is pressed to change the view; this seems not to apply in all views but in most. When an event is fired (i.e. `AUTOBRAKE_HI_SET`), this event sets the LVAR `OVHD_AUTOBRK_MAX_ON_IS_PRESSED`) to `3` for at least 250 ms. That means the Autobrake system is explicitly set to MAX (when allowed) with every event fired. This does not allow the usage of the buttons in the cockpit if those events are fired. In the case of the TCA, the hardware button needs to be put in position `DISARM`, then the buttons can be operated as usual (because in that case no events are fired).

🟢 using the right mouse button has no influence on the Autobrake state
🟢 when using FSUIPC or SPAD.neXt the user can either fire events permanently or using single events; both will work
🟢 when TCA is set to `DISARM` the cockpit buttons can be operated as usual
🟢 default events are explicit, they do what they say instead of activating/deactivating a mode
🟡 buttons cannot be operated when hardware button is set to a certain position stopping hardware and state to get out of sync

aguther commented:
It turned out that when using the TCA with default MSFS binding, during landing it can cause issues when applying manual brakes. It would be impossible to deactivate the autobrake system because it immediately gets activated again.

So I removed the change of the default events and added more events that give more possibilities which should be hopefully fine for everyone.

C6-Skypat commented:
hello

on TCA (was working fine before your code 72 hours ago)
now:
DISARM to BTV to LO = KO -> BRAKE LOW is NOT activated
LO to 2 = OK -> BRAKE MEDIUM is activated
2 to 3 to HI = OK -> BRAKE MAX is activated
HI to 3 to 2 = OK -> BRAKE MEDIUM is activated
2 to LO = OK -> BRAKE LOW is activated
LO to BTV to DISARM = OK -> Brake is disarmed

the issue is DISARM to LO, it don't work at all.

aguther commented:
> hello
> 
> on TCA (was working fine before your code 72 hours ago with default MSFS TCA profile)
> 
> now: DISARM to BTV to LO = KO -> BRAKE LOW is NOT activated LO to 2 = OK -> BRAKE MEDIUM is activated 2 to 3 to HI = OK -> BRAKE MAX is activated HI to 3 to 2 = OK -> BRAKE MEDIUM is activated 2 to LO = OK -> BRAKE LOW is activated LO to BTV to DISARM = OK -> Brake is disarmed
> 
> the issue is DISARM to LO, it don't work at all.

The default events are untouched with the latest code on this PR.

C6-Skypat commented:
> The default events (that are used by the default binding of MSFS) are **untouched** with the latest code on this PR.
> 
> Which version did you try and updated when?
> 
> I will retry with the TCA and default key bindings to be sure and I will try also with the development version.
> 
> Keep in mind that it is also depending on how fast you change your hardware buttons. This is an issue with the approach that is taken up to now and is still in. The time delay is 1.5 s. So when you switch from one position to the other and return quite fast it can omit the event. This is not a problem this PR introduced, it is a problem of the approach as is. But due to how MSFS handles the TCA, there is simply no better way.
> 
> This is the reason why I added additional events that allow a distinct setting and better handling.

hope it will help, i just tested both DEV and EXP build
it is working perfectly with *0b5b201* DEV build, NO issue at all
it is NOT working since an EXP build after last friday and it is NOT working with *c7fabcc* EXP build 

today 17:10 another used said:
_"On my side, the previous var I was using for autobrakes was A32NX_AUTOBRAKES_ARMED_MODE (set via streamdeck/mobiflight). It is not working anymore, I assume after the pull request mentioned above. I also have the TCA and the autobrake bindings via msfs also stopped working"_

best regards, hope you will find.

aguther commented:
Ok, the issue should be fixed now.

C6-Skypat commented:
> Ok, the issue should be fixed now.

yes it's fixed, works again perfectly with TCA.
thx for your works

crocket63 commented on src/systems/a320_systems/src/hydraulic/mod.rs:
```diff
@@ -2381,9 +2383,10 @@ impl A320RamAirTurbineController {
-            && solenoid_1_should_trigger_deployment_if_powered)
-            || (self.is_solenoid_2_powered && solenoid_2_should_trigger_deployment_if_powered);
+        // due to initialization issues the RAT will not deployed in any case when simulation has just started
+        self.should_deploy = (context.simulation_time() >= 2.)
```
Might be a nitpick but couldn't we make that "context.simulation_time() >= 2"  a method like context.is_sim_initialiazing() ?
So any function that just need a simple disable on iinit can use this, and if something changes in the sim and the 2 becomes a 4 seconds, we can just change this at one place.

Saschl replied:
For what it's worth, in JS instruments one can query the GameState which maybe can be used to set the LVAR (e.g. https://github.com/flybywiresim/a32nx/blob/56c798eb25c60ab66500f2a6023333adff6c70d4/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js#L264)

aguther replied:
Thank you for the tips, I think it's much better now!

tracernz commented on src/fmgc/src/components/ReadySignal.ts:
```diff
@@ -0,0 +1,22 @@
+    update(deltaTime: number): void {
+        if (this.updateThrottler.canUpdate(deltaTime) !== -1
+            && this.baseInstrument.getGameState() === GameState.ingame
+            && SimVar.GetSimVarValue('L:A32NX_IS_READY', 'number') !== 1) {
```
We could probably cache the set state in a member variable to save reading the A32NX_IS_READY simvar every frame?

aguther replied:
Don’t forget there is an update throttler above…

tracernz replied:
My bad, I did miss that! This is fine then. 👍

tracernz reviewed (approved):
Looks fine to me. Would be good to get a rusty boy to have a look.

aguther commented on src/systems/systems/src/shared/mod.rs:
```diff
@@ -323,6 +323,18 @@ impl DelayedTrueLogicGate {
         }
     }
 
+    pub fn new_with_init_state(delay: Duration, initial_state: bool) -> DelayedTrueLogicGate {
```
It is necessary to properly initialize this type of gate when in flight.

Let's remember how the gate works (with delay of 4 s):
- t: 0, in: false, out: false
- t: 1, in: true, out: false
- t: 2, in: true, out: false
[...]
- t: 5, in: true, out: true

Now if we spawn in flight, the simulation starts, and the value could be true right from the start. But the gate would always need the delay time to get true. Now this is a generic module, we don't know how everyone is using it. The delay could be small but in theory it could also be 5 minutes. Depending on the usage, there needs to be a way to initialize this gate in a fully true state from the start.

crocket63 replied:
OR: I just totally misunderstood the use of this :D   Let me have another look I'm tired :p

aguther replied:
The idea is that you can fundamentally initialize the gate in a true/true state. Once it goes to false shortly, the normal behavior should happen.

crocket63 replied:
Ok! So would it make sense to rename this to something like bypass_initial_delay ?

My idea behind this is this way the value you give there will not depend anymore on the type of gate, as a DelayedFalse you would still ask to "bypass initial delay = true"

aguther replied:
Yes...David made a good proposal for an improvement that I will use. By doing this, it gets much better understandable.

crocket63 replied:
Nice :)

aguther commented on src/systems/a320_systems/src/hydraulic/mod.rs:
```diff
@@ -3285,16 +3290,18 @@ impl A320AutobrakeController {
             target: Acceleration::new::<meter_per_second_squared>(0.),
             mode: AutobrakeMode::NONE,
-            arming_is_allowed_by_bcu: false,
+            arming_is_allowed_by_bcu: context.is_in_flight(),
```
From a logical point of view, I consider this to be more correct because it can be assumed that in flight the BCU would allow this.

davidwalschots reviewed (commented):
The various changes in `src/systems/a320_systems/src/hydraulic/mod.rs` can be tested automatically. I'd recommend doing so as per `src/systems/guidelines.md`.
